Python/Product/Analysis/LanguageServer/Server.cs

```diff
@@ -1022,71 +988,79 @@ public sealed class Server : ServerBase, IDisposable {
             }
         }
 
-        private async Task EnqueueItem(IDocument doc, AnalysisPriority priority = AnalysisPriority.Normal, bool enqueueForAnalysis = true) {
-            try {
-                VersionCookie vc;
-                using (_pendingAnalysisEnqueue.Incremented()) {
-                    IAnalysisCookie cookie;
-                    using (GetDocumentParseCounter(doc, out int count)) {
-                        if (count > 3) {
-                            // Rough check to prevent unbounded queueing. If we have
-                            // multiple parses in queue, we will get the latest doc
-                            // version in one of the ones to come.
-                            return;
-                        }
-
-                        TraceMessage($"Parsing document {doc.DocumentUri}");
-                        cookie = await _parseQueue.Enqueue(doc, _analyzer.LanguageVersion).ConfigureAwait(false);
+        private void EnqueueItem(IDocument doc, AnalysisPriority priority = AnalysisPriority.Normal, bool enqueueForAnalysis = true) {
+            VersionCookie vc;
+            using (_pendingAnalysisEnqueue.Incremented()) {
+                Task<IAnalysisCookie> cookieTask;
+                using (GetDocumentParseCounter(doc, out var count)) {
+                    if (count > 3) {
+                        // Rough check to prevent unbounded queueing. If we have
+                        // multiple parses in queue, we will get the latest doc
+                        // version in one of the ones to come.
+                        return;
                     }
 
-                    vc = cookie as VersionCookie;
-                    if (vc != null) {
-                        foreach (var kv in vc.GetAllParts(doc.DocumentUri)) {
-                            ParseComplete(kv.Key, kv.Value.Version);
-                        }
-                    } else {
-                        ParseComplete(doc.DocumentUri, 0);
-                    }
+                    TraceMessage($"Parsing document {doc.DocumentUri}");
+                    cookieTask = _parseQueue.Enqueue(doc, _analyzer.LanguageVersion);
+                }
 
-                    if (doc is IAnalyzable analyzable && enqueueForAnalysis) {
-                        TraceMessage($"Enqueing document {doc.DocumentUri} for analysis");
-                        _queue.Enqueue(analyzable, priority);
+                // The call must be fire and forget, but should not be yielding.
+                // It is called from DidChangeTextDocument which must fully finish
+                // since otherwise Complete() may come before the change is enqueued
+                // for processing and the completion list will be driven off the stale data.
+                cookieTask.ContinueWith(t => {
```

Comment:
For the sake of indentation, can we extract this lambda to its own function ("FinishParse" or similar)? At least for the non-faulting case.

Reply:
Done